### PR TITLE
Tweak some algorithm naming inconsistencies

### PR DESCRIPTION
**Pull request by mattcaswell**
Also update the documentation and include a description of the "canonical" naming approach.

### CODE REVIEW

File: doc/man7/provider.pod [hunk 1]
```diff
@@ -224,6 +216,28 @@ In this case an algorithm implementation is implicitly fetched using
 default search criteria and an algorithm name that is consistent with
 the type of EVP_PKEY being used.
 
+=head3 Algorithm naming
```
Comment (levitte) [inline doc/man7/provider.pod]: This is not the best place for this description, as providers are by design capable of supplying more than just cryptographic algorithms.  For example, `OSSL_STORE` implementations are named by URI scheme.

I would suggest a separate file for this.  Still in man7/...  or possibly in internal/man7/, depending on if you think this is public or internal info.

File: doc/man7/provider.pod [hunk 2]
```diff
@@ -276,13 +323,24 @@ It currently offers the following FIPS approved named algorithms:
 
 =item Digests
 
-SHA1, SHA224, SHA256, SHA384, SHA512, SHA512-224, SHA512-256,
```
Comment (t8m) [inline doc/man7/provider.pod]: Specifying all the algorithms a provider supports duplicates the information so it potentially presents a maintenance problem. Also it conflicts with #10088 unless the manual page would be generated from a template. That would be one option. Another option would be to just reference an openssl provider -list-algos command or such when we will have it.

Comment (levitte) [reply doc/man7/provider.pod]: Errrrrrr...  I dunno how it conflicts with #10088.  It's not like I'm making any implementation more or less available.  However, it might be good to note that the availability is subject to configuration and that for a definitive list relevant to the local installation, something like `openssl list -digest-algorithms` et al should be used.

Comment (mattcaswell) [reply doc/man7/provider.pod]: I kind of agree having just spent the time to update this list, that this is going to become a bit of a pain, so I'll replace it with a reference to `openssl list`. However I don't think that can do the asym algorithms yet (partly I guess because we are only part way through converting the asym stuff to providers).

File: doc/man7/provider.pod (hunk 1 above)
Comment (mattcaswell) [reply doc/man7/provider.pod]: I'm not sure why you think this isn't a good place for it. I understand that an identifier in a fetch can be something other than an algorithm name. The page is quite explicit on that point:

````
This is most commonly an algorithm name (this is the case for all EVP
methods), but may also be called something else.
````

But, I don't see why that prevents us from talking about algorithm names here (especially as this is "most commonly" the case). I chose the title "algorithm naming" on purpose because of this issue, rather than "identifier naming".

**Comment by richsalz** [comment]
It doesn't seem worth doing, but I thought I'd ask anyway:  worth adding to find-doc-nits?

File: doc/man7/provider.pod (hunk 2 above)
Comment (levitte) [reply doc/man7/provider.pod]: Yup...  I've code somewhere that expands what `openssl list` as well as `openssl provider`can display.

**Comment by mattcaswell** [comment]
> It doesn't seem worth doing, but I thought I'd ask anyway: worth adding to find-doc-nits?

Hmm. I'm not sure. It may be tricky because I think there are some places in the docs which are legitimately referring to legacy names, because of legacy APIs. It's probably not worth it at this stage. Maybe in the future.

**Comment by levitte** [comment]
I'd very much like to know how much AI we're gonna put in find-doc-nits :wink:
Or perhaps I misunderstand what @richsalz wants to have i check...

**Comment by mattcaswell** [comment]
> I'd very much like to know how much AI we're gonna put in find-doc-nits

I'd quite like it to automatically write the documentation for us. I'm sure your perl skills are up to it @levitte!

**Comment by mattcaswell** [comment]
Fixup pushed addressing the comment from @t8m.

File: doc/man7/provider.pod (hunk 2 above)
Comment (richsalz) [reply doc/man7/provider.pod]: Several manpages list algorithms -- and TLS versions.  Which ones are actually possible depends on the configuration, and a recent PR added the "ifdef" markup so that find-doc-nits wouldn't complain. Most of the pages, and the online help, say "any supported cipher" and the docs reference the openssl-cipher page.

I think having some typical options is worthwhile because it helps new users understand things better. The difference between AES, AES-GCM, AES-keywrap is not obvious, for example.

TL;DR (which should come at the top): you'll never satisfy all constraints.

File: providers/default/defltprov.c [hunk 3]
```diff
@@ -73,6 +73,9 @@ static int deflt_get_params(const OSSL_PROVIDER *prov, OSSL_PARAM params[])
  * We add diverse other names where applicable, such as the names that
  * NIST uses, or that are used for ASN.1 OBJECT IDENTIFIERs, or names
  * we have used historically.
+ *
+ * Algorithm names are case insenstive, but we use all caps in our "canoncial"
```
Comment (t8m) [inline providers/default/defltprov.c]: insensitive, canonical typos

**Comment by slontis** [comment]
Apart from a rebase, and fixing a typo in defltprov.c this looks ready to go..

**Comment by mattcaswell** [comment]
Sorrry, this got forgotten about. Fixup pushed addressing typo nits.

@t8m (or anyone) please take a look.

**Comment by mattcaswell** [comment]
Ping @openssl/committers

File: doc/man7/provider.pod [hunk 4]
```diff
@@ -216,42 +216,45 @@ In this case an algorithm implementation is implicitly fetched using
...
-AES-256-CFB8, AES-192-CFB8, AES-128-CFB8, AES-256-CTR, AES-192-CTR,
-AES-128-CTR, id-aes256-GCM, id-aes192-GCM, id-aes128-GCM
+OpenSSL comes with a set of providers.
 
-=item Key Exchange
+The algorithms available in each of these providers may vary due to build time
+configuration options. The L<openssl-list(1)> command can be used to list the
+currently available algorithms.
+
+The algorithms names shown from L<openssl-list(1)> can be used as an algorithm
```
Comment (paulidale) [inline doc/man7/provider.pod]: Either _algorithm names_ or _algorithms' names_ here.  The former reads better but loses the plurality.

Maybe _The names of the algorithms shown from_ is the best by spelling out the genitive directly.

File: doc/man7/provider.pod [hunk 5]
```diff
@@ -216,42 +216,45 @@ In this case an algorithm implementation is implicitly fetched using
...
+identifier to the appropriate fetching function.
 
-dhKeyAgreement
+=head2 Default provider
 
-=back
+The default provider is built in as part of the F<libcrypto> library.
+Should it be needed (if other providers are loaded and offer
```
Comment (paulidale) [inline doc/man7/provider.pod]: _mandated_ instead of _needed_ ? Perhaps?

**Comment by mattcaswell** [comment]
Pushed. Thanks. I took your first wording suggestion, but left the other one as it was.
